File: Data Structures & Algorithms/same-binary-tree/submission-5.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} p
     * @param {TreeNode} q
     * @return {boolean}
     */
    isSameTree(p: TreeNode | null, q: TreeNode | null): boolean {

        const dfs = (nodeP: TreeNode, nodeQ) => {
            if (!nodeP && !nodeQ) return true // means both are not null
            if (!nodeP || !nodeQ) return false // both are not null but one is, in which case not equal

            if (nodeP.val !== nodeQ.val) return false

            const left = dfs(nodeP.left, nodeQ.left)
            if (!left) return false

            const right = dfs(nodeP.right, nodeQ.right)
            if (!right) return false

            return true
        }

        return dfs(p, q)
    }
}

File: Data Structures & Algorithms/climbing-stairs/submission-5.py
class Solution:
    def climbStairs(self, n: int) -> int:
        arr = [None] * (n + 1)

        def dfs(i):
            if i == n:
               return 1

            if i > n:
                return 0
            
            if arr[i] is not None:
                print(i, arr)
                return arr[i]

            left = dfs(i + 1)
            right = dfs(i + 2)

            arr[i] = left + right

            return arr[i]

        return dfs(0) 